diff -Naur nginx-1.2.4/src/http/modules/ngx_http_ssl_module.c nginx-1.2.4-tack/src/http/modules/ngx_http_ssl_module.c
--- nginx-1.2.4/src/http/modules/ngx_http_ssl_module.c	2012-01-18 19:07:43.000000000 +0400
+++ nginx-1.2.4-tack/src/http/modules/ngx_http_ssl_module.c	2012-10-09 18:14:37.458463105 +0400
@@ -152,6 +152,31 @@
       offsetof(ngx_http_ssl_srv_conf_t, crl),
       NULL },
 
+#ifndef OPENSSL_NO_TACK
+
+    { ngx_string("ssl_tack"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_flag_slot,
+      NGX_HTTP_SRV_CONF_OFFSET,
+      offsetof(ngx_http_ssl_srv_conf_t, tack),
+      NULL },
+
+    { ngx_string("ssl_tack_file"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_HTTP_SRV_CONF_OFFSET,
+      offsetof(ngx_http_ssl_srv_conf_t, tack_file),
+      NULL },
+
+    { ngx_string("ssl_tack_pin_activation"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_CONF_FLAG,
+      ngx_conf_set_flag_slot,
+      NGX_HTTP_SRV_CONF_OFFSET,
+      offsetof(ngx_http_ssl_srv_conf_t, tack_pin_activation),
+      NULL },
+
+#endif
+
       ngx_null_command
 };
 
@@ -337,6 +362,11 @@
     sscf->builtin_session_cache = NGX_CONF_UNSET;
     sscf->session_timeout = NGX_CONF_UNSET;
 
+#ifndef OPENSSL_NO_TACK
+    sscf->tack = NGX_CONF_UNSET;
+    sscf->tack_pin_activation = NGX_CONF_UNSET;
+#endif
+
     return sscf;
 }
 
@@ -387,6 +417,11 @@
 
     ngx_conf_merge_str_value(conf->ciphers, prev->ciphers, NGX_DEFAULT_CIPHERS);
 
+#ifndef OPENSSL_NO_TACK
+    ngx_conf_merge_value(conf->tack, prev->tack, 0);
+    ngx_conf_merge_str_value(conf->tack_file, prev->tack_file, "");
+    ngx_conf_merge_value(conf->tack_pin_activation, prev->tack_pin_activation, 0);
+#endif
 
     conf->ssl.log = cf->log;
 
@@ -515,6 +550,26 @@
         return NGX_CONF_ERROR;
     }
 
+#ifndef OPENSSL_NO_TACK
+
+    if (conf->tack && conf->tack_file.len) {
+        if (ngx_conf_full_name(cf->cycle, &conf->tack_file, 1) != NGX_OK) {
+            return NGX_CONF_ERROR;
+        }
+        if (SSL_CTX_use_tack_file(conf->ssl.ctx,
+                                    (const char *) conf->tack_file.data,
+                                    conf->tack_pin_activation)
+            == 0)
+        {
+            ngx_ssl_error(NGX_LOG_EMERG, cf->log, 0,
+                            "Unable to initialize TLS TACK extension "
+                            "(incompatible OpenSSL version?)");
+            return NGX_CONF_ERROR;
+        }
+    }
+
+#endif
+
     return NGX_CONF_OK;
 }
 
diff -Naur nginx-1.2.4/src/http/modules/ngx_http_ssl_module.h nginx-1.2.4-tack/src/http/modules/ngx_http_ssl_module.h
--- nginx-1.2.4/src/http/modules/ngx_http_ssl_module.h	2012-01-18 19:07:43.000000000 +0400
+++ nginx-1.2.4-tack/src/http/modules/ngx_http_ssl_module.h	2012-10-09 18:14:58.410199578 +0400
@@ -43,6 +43,13 @@
 
     u_char                         *file;
     ngx_uint_t                      line;
+
+#ifndef OPENSSL_NO_TACK
+    ngx_flag_t                      tack;
+    ngx_str_t                       tack_file;
+    ngx_flag_t                      tack_pin_activation;
+#endif
+
 } ngx_http_ssl_srv_conf_t;
 
 
